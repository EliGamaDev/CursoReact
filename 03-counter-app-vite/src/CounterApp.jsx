import {useState} from 'react'
import PropTypes from 'prop-types'

export const CounterApp=({value})=>{

    //Si no interactúa con el 'value', puede quedar fuera del componente para que no
    //ocupe espacio en memoria siempre que se vuelva a renderizar...
    //OJO: Cuando cambia el estado (render), el componente SE VUELVE A EJECUTAR
    //console.log(`Render: ${value}`)
    
    //const [counter,setCounter]=useState(0);
    const [counter,setCounter]=useState(value);
    
    const handleAdd=(event)=>{
        //console.log(event);
        //console.log(newValue)

        //SetCounter: El estado cambió, re ejecuta la renderización del componente...
        //setCounter(counter+1) 
        setCounter((c)=>c+1)

    }

    const handleSus=(event)=>{
        setCounter((c)=>c-1)
    }

    const handleReset=(event)=>{
        setCounter((c)=>value)
    }

    //HOOKS: Funciones
    
    return(
        <>
            <h1>CounterApp</h1>
            <h2> {counter} </h2>
            
            <button onClick={handleAdd}>+1</button>
            <button onClick={handleSus}>-1</button>
            <button aria-label="btn-reset" onClick={handleReset}>Reset</button>
            
            {/*<input type="text"></input>*/}
        </>
    )
}

CounterApp.propTypes={
    value:PropTypes.number.isRequired
}

CounterApp.defaultProps={
    value:0
} 