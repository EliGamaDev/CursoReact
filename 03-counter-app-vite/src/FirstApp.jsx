//import { Fragment } from "react"
//<></>: Indica FRAGMENT sin usar el import
import PropTypes from 'prop-types';

const newMsg='Eli';
const array=[1,2,3,4,5,6,7,8,9];
const obj={
    message:'Hola ',
    title:'Eli'
}

const getResultado=(a,b)=>{
    return a+b;
}

//Propiedades -> Properties -> props (es objeto) en argumento...
//Canal de comunicación con el padre: <React.StrictMode>

export const FirstApp=({title,subTitle,name})=>{
    /*return(  //Esto crea un div INNECESARIO, le da más procesamiento a React
        <div>
            <h1>First App de Eli</h1>
            <p>Primer párrafo...</p>
        </div>
    )*/ 

    /*if(!title){
        throw new Error('ERROR: title vacío...');
    }*/

    return(
        //<Fragment>
            //<h1>First App de Eli</h1>
            //<p>Primer párrafo...</p>
        //</Fragment>
        
        <>
            <h1 data-testid="test-title">{title}</h1>
            <h2>{subTitle}</h2>
            <h2>{subTitle}</h2>
            <h2>{name}</h2>
        </>
            /*Meter en <></>:
            {<h1>First App de {newMsg}</h1>
            <h2>ARRAY: {array}</h2>
            <h3>OBJETO: {obj.message+' '+obj.title}</h3>
            <code>OBJETO JSON: {JSON.stringify(obj)}</code>
            <p>Primer párrafo...</p>
            <br></br>
            <h3>Resultado: {getResultado(5,4)}</h3>}*/

            
        
    )
}

FirstApp.propTypes={
    title:PropTypes.string.isRequired,
    //subTitle:PropTypes.number.isRequired
    subTitle:PropTypes.string.isRequired
}

//Aplica propiedades por defecto... Anula mensajes error.
FirstApp.defaultProps={
    //title:'No hay título...',
    subTitle:'1234',
    name:'name: Eli'
}