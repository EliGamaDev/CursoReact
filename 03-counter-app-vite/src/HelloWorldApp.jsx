import React from 'react'

//export const HelloWorldApp=()=>(<h1>Arrow Function</h1>)

/*export function App(){
    //document.createElement();
    return <h1>Hola Mundo</h1>;
}*/

export const HelloWorldApp = () => {
  return (
    <h1>Hello World App</h1>
    //<div>Hello World App</div>
  )
}
