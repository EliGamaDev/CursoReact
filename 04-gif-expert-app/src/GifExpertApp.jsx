import {useState} from 'react'
import { AddCategory } from './components/AddCategory';

export const GifExpertApp=()=>{

    //Hook de React para mantener ESTADO: useState
    const [categories, setCategories] = useState(['Pokémon','Dragon Ball']);

    const onAddCategory=()=>{

        //console.log('Agregar Categoría...');
        console.log(categories);
        //categories.push('Naruto');
        //setCategories((c)=>c+'Naruto');

        setCategories([...categories,'Naruto']);

        console.log(categories);
        //return categories;

    }
    
    return(
        <>
            {/* titulo */}
            <h1>GifExpertApp</h1>

            {/* input */}
            <AddCategory setCategories={setCategories}/>

            {/* listado de gif */}
            <button onClick={onAddCategory}>Agregar Categoría</button>
            <ol>
                { categories.map(category=>{
                    return <li key={ category }>{category}</li>
                }) }
                {/*<li></li>*/}
            </ol>
                {/* gif item */}
        </>
    )

}