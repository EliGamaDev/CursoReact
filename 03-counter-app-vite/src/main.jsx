//JSX: Código REACT: JS + XML
import React from 'react';
import ReactDOM from 'react-dom/client';
//import {HelloWorldApp} from './HelloWorldApp'
//import {FirstApp} from './FirstApp'
import {CounterApp} from './CounterApp'

//Estilo global CounterApp:
import './styles.css'

//Renderizar aplicación en DOM...
ReactDOM.createRoot(document.getElementById('root')).render(
    <React.StrictMode>
        <CounterApp value={20}/>
    </React.StrictMode>
)

/*<FirstApp title="Hola, soy Eli..." />
        
    <HelloWorldApp ARRIBA: 'title' es PROPS />  
    title="Hola, soy Eli 2"
    subTitle={12345}
        
*/

/*

PRUEBAS: AAA -> Arrange (arreglar), Act (actuar) y Assert (afirmar)...

- Arrange: Preparamos el estado inicial: Iniciamos variables, importaciones necesarias...
- Act: Aplicamos acciones o estímulos al sujeto de pruebas: Llamar métodos, simular clicks, realizar
     acciones sobre el paso anterior... 
- Assert: Observar el comportamiento resultante: son los resultados esperados?

MITOS: 

- Hacen que mi app no tenga errores: X
- Las pruebas no pueden fallar: X
- Hacen más lenta mi aplicación: X
- Es una pérdida de tiempo: X
- Hay que probar todo: X

*/