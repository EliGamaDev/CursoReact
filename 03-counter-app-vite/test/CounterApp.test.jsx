/** @jest-environment jsdom */

import { render,screen,fireEvent } from "@testing-library/react";
import { CounterApp } from "../src/CounterApp"
//Tarea: Pruebas en <CounterApp />

//1. Debe hacer match con el Snapshot
//2. Debe mostrar el valor inicial de 100.

describe('Pruebas en <CounterApp />',()=>{

    const valorInicial=10;

    test('Prueba 3.1: Debe hacer match con el Snapshot',()=>{

        const {container}=render(<CounterApp value={valorInicial}/>);
        expect(container).toMatchSnapshot();

    })

    test('Prueba 3.2: Debe mostrar el valor incial de 100',()=>{

        render(<CounterApp value={100}/>);

        expect(screen.getByText('100')).toBeTruthy();
        //expect(screen.getByRole('heading',{level:2}).innerHTML).toContain(valorInicial);

        //expect(screen.getByRole('heading',{level:2}).innerHTML).toContain(valorInicial);


    })

    test('Prueba 3.3: Simular click incrementador +1',()=>{

        render(<CounterApp value={valorInicial}/>);
        //fireEvent.click(screen.getByText('+1'));
        fireEvent.click(screen.getByText('+1'));
        expect(screen.getByText('11')).toBeTruthy();

    })

    test('Prueba 3.4: Simular click decrementador -1',()=>{

        render(<CounterApp value={valorInicial}/>);
        //fireEvent.click(screen.getByText('+1'));
        fireEvent.click(screen.getByText('-1'));
        
        expect(screen.getByText('9')).toBeTruthy();

    })

    test('Prueba 3.5: Simular click botÃ³n "RESET"',()=>{

        render(<CounterApp value={valorInicial}/>);
        //screen.debug();
        fireEvent.click(screen.getByText('+1'));
        fireEvent.click(screen.getByText('+1'));
        fireEvent.click(screen.getByText('+1'));
        //screen.debug();
        //fireEvent.click(screen.getByText('Reset'));
        //screen.debug();

        fireEvent.click(screen.getByRole('button',{name:'btn-reset'}));

        expect(screen.getAllByText(valorInicial)).toBeTruthy();

    })

})