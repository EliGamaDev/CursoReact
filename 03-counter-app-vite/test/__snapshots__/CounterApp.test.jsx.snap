// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <CounterApp /> Prueba 3.1: Debe hacer match con el Snapshot 1`] = `
<div>
  <h1>
    CounterApp
  </h1>
  <h2>
     
    10
     
  </h2>
  <button>
    +1
  </button>
  <button>
    -1
  </button>
  <button
    aria-label="btn-reset"
  >
    Reset
  </button>
</div>
`;
